python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,267 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility module for various conversions around PrimitiveType implementations
+
+This module enables:
+    - Converting partition strings to built-in python objects
+    - Converting a value to a byte buffer
+    - Converting a byte buffer to a value
+
+Note:
+    Conversion logic varies based on the PrimitiveType implementation. Therefore conversion functions
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+
+Example (registering `from_partition_to_py` logic for BooleanType):
+    >>> @from_partition_value_to_py.register(BooleanType)
+    >>> def _(primitive_type, partition_value_as_str: str):
+            if partition_value_as_str is None:
+                return False
+            return partition_value_as_str.lower() == "true"
+
+Example (stacking registrations):
+   >>> @from_partition_value_to_py.register(FloatType)
+   >>> @from_partition_value_to_py.register(DoubleType)
+   >>> def _(primitive_type, partition_value_as_str: str):
+           return float(partition_value_as_str)
+"""
+import struct
+import uuid
+from decimal import ROUND_HALF_DOWN, Decimal, localcontext
+from functools import singledispatch
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+@singledispatch
+def from_partition_value_to_py(primitive_type, partition_value_as_str: str):
+    """A generic function function which converts a partition string to a python built-in
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        partition_value_as_str(str): A string representation of a partition value
+    """
+    if partition_value_as_str is None:
+        return None
+    raise TypeError(
+        f"Cannot convert partition string to python built-in, type {primitive_type} not supported: '{partition_value_as_str}'"
+    )
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return False
```

Comment:
In this case since it is very constrained values it could be but the general recommendation is to use [direct comparisons with None](https://peps.python.org/pep-0008/#programming-recommendations) if that is actually your intent.

Reply:
I made an assumption here that `None` is all we need to explicitly check for. Thinking about this more for the other functions, how do we want to handle a `StringType()` or `IntegerType()` partition that has a partition value of an empty string? Should that also return `None`?